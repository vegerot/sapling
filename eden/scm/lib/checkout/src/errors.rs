/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#[derive(Debug, thiserror::Error)]
#[error("checkout error: {source}")]
pub struct CheckoutError {
    pub resumable: bool,
    pub source: anyhow::Error,
}

#[derive(Debug, thiserror::Error)]
#[error("error updating {path}: {message}")]
pub struct EdenConflictError {
    pub path: String,
    pub message: String,
}
