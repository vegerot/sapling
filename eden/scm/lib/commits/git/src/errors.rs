/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use dag::Vertex;

pub fn hash_mismatch(a: &Vertex, b: &Vertex) -> anyhow::Error {
    anyhow::format_err!("hash mismatch: {:?} != {:?}", a, b)
}
