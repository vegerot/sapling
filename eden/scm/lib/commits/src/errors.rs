/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use dag::Vertex;

pub fn test_only(name: &str) -> anyhow::Error {
    anyhow::format_err!("{} should only be used in tests", name)
}

pub fn hash_mismatch(a: &Vertex, b: &Vertex) -> anyhow::Error {
    anyhow::format_err!("hash mismatch: {:?} != {:?}", a, b)
}
