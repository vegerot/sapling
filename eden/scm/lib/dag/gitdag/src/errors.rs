/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

pub trait MapDagError<T> {
    fn context(self, message: &'static str) -> dag::Result<T>;
    fn with_context(self, func: impl Fn() -> String) -> dag::Result<T>;
}

impl<T> MapDagError<T> for Result<T, git2::Error> {
    fn context(self, message: &'static str) -> dag::Result<T> {
        anyhow::Context::context(self, message)
            .map_err(|e| dag::errors::BackendError::Other(e).into())
    }
    fn with_context(self, func: impl Fn() -> String) -> dag::Result<T> {
        anyhow::Context::with_context(self, func)
            .map_err(|e| dag::errors::BackendError::Other(e).into())
    }
}
