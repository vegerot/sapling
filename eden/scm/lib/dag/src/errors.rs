/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::io;
use std::num::TryFromIntError;

use thiserror::Error;

use crate::Group;
use crate::Id;
use crate::Vertex;

/// Error used by the Dag crate.
#[derive(Debug, Error)]
pub enum DagError {
    /// A vertex name cannot be found.
    #[error("{0:?} cannot be found")]
    VertexNotFound(Vertex),

    /// An Id cannot be found.
    #[error("{0:?} cannot be found")]
    IdNotFound(Id),

    /// A fast path cannot be used.
    #[error("NeedSlowPath: {0}")]
    NeedSlowPath(String),

    /// Callsite does something wrong. For example, a "parent function" does not
    /// return reproducible results for a same vertex if called twice.
    #[error("ProgrammingError: {0}")]
    Programming(String),

    /// Logic error in this crate. A bug in this crate or the backend data.
    #[error("bug: {0}")]
    Bug(String),

    /// The backend (ex. filesystem) cannot fulfill the request somehow.
    #[error(transparent)]
    Backend(Box<BackendError>),

    /// No space for new Ids.
    #[error("out of space for group {0:?}")]
    IdOverflow(Group),

    /// Integer conversion overflow.
    #[error(transparent)]
    IntOverflow(#[from] TryFromIntError),
}

#[derive(Debug, Error)]
pub enum BackendError {
    #[error("{0}")]
    Generic(String),

    #[error(transparent)]
    Io(#[from] std::io::Error),

    #[cfg(any(test, feature = "indexedlog-backend"))]
    #[error(transparent)]
    IndexedLog(#[from] indexedlog::Error),

    /// Other source of backend errors. Useful for external crates implementing
    /// traits of the `dag` crate.
    #[error(transparent)]
    Other(#[from] anyhow::Error),
}

impl From<BackendError> for DagError {
    fn from(err: BackendError) -> DagError {
        DagError::Backend(Box::new(err))
    }
}

#[cfg(any(test, feature = "indexedlog-backend"))]
impl From<indexedlog::Error> for DagError {
    fn from(err: indexedlog::Error) -> DagError {
        DagError::Backend(Box::new(BackendError::from(err)))
    }
}

impl From<io::Error> for DagError {
    fn from(err: io::Error) -> DagError {
        DagError::Backend(Box::new(BackendError::from(err)))
    }
}

/// Quick way to return a `BackendError::Generic` error.
pub fn bug<T>(message: impl ToString) -> crate::Result<T> {
    Err(DagError::Bug(message.to_string()))
}

/// Quick way to return a `Programming` error.
pub fn programming<T>(message: impl ToString) -> crate::Result<T> {
    Err(DagError::Programming(message.to_string()))
}

pub trait NotFoundError {
    fn not_found_error(&self) -> DagError;

    fn not_found<T>(&self) -> crate::Result<T> {
        Err(self.not_found_error())
    }
}

impl NotFoundError for Id {
    fn not_found_error(&self) -> DagError {
        ::fail::fail_point!("dag-not-found-id");
        DagError::IdNotFound(self.clone())
    }
}

impl NotFoundError for Vertex {
    fn not_found_error(&self) -> DagError {
        ::fail::fail_point!("dag-not-found-vertex");
        DagError::VertexNotFound(self.clone())
    }
}
