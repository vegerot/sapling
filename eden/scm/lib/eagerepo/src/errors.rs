/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use dag::Vertex;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum Error {
    #[error(transparent)]
    Dag(#[from] dag::Error),

    #[error("hash mismatch ({0:?} != {1:?})")]
    HashMismatch(Vertex, Vertex),

    #[error(
        "EagerRepo detected unsupported requires at {0}:\n  Unsupported: {1:?}\n  Missing: {2:?}\n(is there a non-EagerRepo created accidentally at the same location?)"
    )]
    RequirementsMismatch(String, Vec<String>, Vec<String>),

    #[error(
        "when adding commit {0:?} with root tree {1:?}, referenced paths {2:?} are not present"
    )]
    CommitMissingPaths(Vertex, Vertex, Vec<String>),

    #[error("when moving bookmark {0:?} to {1:?}, the commit does not exist")]
    BookmarkMissingCommit(String, Vertex),

    #[error(transparent)]
    Other(#[from] anyhow::Error),
}

impl From<std::io::Error> for Error {
    fn from(err: std::io::Error) -> Self {
        Self::Dag(dag::errors::BackendError::from(err).into())
    }
}

impl From<zstore::Error> for Error {
    fn from(err: zstore::Error) -> Self {
        anyhow::Error::from(err).into()
    }
}

impl From<metalog::Error> for Error {
    fn from(err: metalog::Error) -> Self {
        anyhow::Error::from(err).into()
    }
}
