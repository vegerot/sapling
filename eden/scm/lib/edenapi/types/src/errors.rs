/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use serde::Deserialize;
use serde::Serialize;
use thiserror::Error;

#[derive(Clone, Debug, Error, Eq, PartialEq)]
#[derive(Serialize, Deserialize)] // used to convert to Python
#[cfg_attr(
    any(test, feature = "for-tests"),
    derive(quickcheck_arbitrary_derive::Arbitrary)
)]
#[error("server error (code {code}): {message}")]
/// Common error structure between Mononoke and Mercurial.
/// The `message` field is self explanatory, a natural language description of the issue that was
/// encountered.
/// The `code` field represents a numeric identifier of the type of issue that was encountered. In
/// most situations the code will be `0`, meaning that there is nothing special about the error.
/// Non-zero codes are used for situations where the client wants to take a specific action (when
/// the client needs to handle that error).
///
/// Error code list:
/// ---------------
/// 1: SegmentedChangelogMismatchedHeads
///    Fatal inconsistency between client and server. The client will want to reclone in this
///    situation.
/// 2: HexError
///    Failed to convert hex to binary hash.
pub struct ServerError {
    pub message: String,
    pub code: u64,
}

impl ServerError {
    pub fn new<M: Into<String>>(m: M, code: u64) -> Self {
        Self {
            message: m.into(),
            code,
        }
    }

    pub fn generic<M: Into<String>>(m: M) -> Self {
        Self::new(m, 0)
    }
}

impl From<types::hash::HexError> for ServerError {
    fn from(e: types::hash::HexError) -> Self {
        Self::new(e.to_string(), 2)
    }
}
