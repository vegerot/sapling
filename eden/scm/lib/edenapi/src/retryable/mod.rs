/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use std::time::Duration;

use async_trait::async_trait;
use edenapi_trait::Entries;
use edenapi_trait::Response;
use futures::channel::mpsc;
use futures::prelude::*;
use http_client::Stats;

use crate::client::Client;
use crate::errors::EdenApiError;

mod files;
mod trees;

pub(crate) use files::RetryableFileAttrs;
pub(crate) use files::RetryableFiles;
pub(crate) use trees::RetryableTrees;

#[async_trait]
pub(crate) trait RetryableStreamRequest: Sized + Sync + Send + 'static {
    type Item: Send + 'static;

    async fn perform(&self, client: Client) -> Result<Response<Self::Item>, EdenApiError>;

    fn received_item(&mut self, _item: &Self::Item) {}

    fn retry_after(
        &mut self,
        error: &EdenApiError,
        attempt: usize,
        max: usize,
    ) -> Option<Duration> {
        if error.is_retryable() && attempt < max {
            // Retrying for a longer period of time is simply a
            // way to wait until whatever surge of traffic is happening ends.
            if error.is_rate_limiting() {
                Some(Duration::from_secs(
                    u32::pow(2, std::cmp::min(3, attempt as u32 + 1)) as u64,
                ))
            } else {
                Some(Duration::from_secs(attempt as u64 + 1))
            }
        } else {
            None
        }
    }

    async fn perform_with_retries(
        self,
        client: Client,
    ) -> Result<Response<Self::Item>, EdenApiError> {
        struct RetryState<R, T> {
            request: R,
            entries: Option<Entries<T>>,
            attempt: usize,
        }

        let state = RetryState {
            request: self,
            entries: None,
            attempt: 0,
        };

        let (stats_tx, stats_rx) = mpsc::channel(0);
        let entries = stream::unfold(state, move |mut state| {
            let mut stats_tx = stats_tx.clone();
            let client = client.clone();

            async move {
                let max_attempts = client.config().max_retry_per_request;
                loop {
                    // Ideally we'd return None when we hit the final error, but we need to use
                    // that time to return the error, so instead we return None on this next
                    // iteration.
                    if state.attempt > max_attempts {
                        return None;
                    }

                    let res = if let Some(ref mut entries) = state.entries {
                        tracing::trace!("Polling response stream");
                        if let Some(res) = entries.next().await {
                            tracing::trace!("Item received");
                            res
                        } else {
                            tracing::trace!("Transfer complete");
                            return None;
                        }
                    } else {
                        tracing::trace!("No active response stream; sending new request");
                        let res = state.request.perform(client.clone()).await;
                        match res {
                            Ok(Response { entries, stats, .. }) => {
                                state.entries = Some(entries);
                                let _ = stats_tx.try_send(stats);
                                continue;
                            }
                            Err(e) => Err(e),
                        }
                    };

                    let error = match res {
                        Ok(item) => {
                            state.request.received_item(&item);
                            return Some((Ok(item), state));
                        }
                        Err(e) => e,
                    };

                    let retry_after =
                        match state
                            .request
                            .retry_after(&error, state.attempt, max_attempts)
                        {
                            Some(d) => d,
                            None => {
                                state.attempt = max_attempts + 1;
                                return Some((Err(error), state));
                            }
                        };
                    state.attempt += 1;
                    state.entries = None;

                    tracing::error!(
                        "Retrying after {:#?} due to error (attempt {}): {}",
                        &retry_after,
                        state.attempt,
                        &error
                    );

                    tokio::time::sleep(retry_after).await;
                }
            }
        })
        .boxed();

        let stats = stats_rx.into_future().then(|(next, _tail)| match next {
            Some(fut) => fut,
            None => Box::pin(future::ok(Stats::default())),
        });

        Ok(Response {
            entries,
            stats: Box::pin(stats),
        })
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashSet;
    use std::sync::Arc;

    use anyhow::anyhow;
    use anyhow::Result;
    use async_runtime::block_on;
    use async_runtime::stream_to_iter;
    use async_trait::async_trait;
    use http_client::HttpClientError;
    use parking_lot::Mutex;
    use types::HgId;
    use types::Key;
    use types::RepoPathBuf;

    use super::*;
    use crate::builder::HttpClientBuilder;
    use crate::EdenApiError;
    use crate::Response;

    pub(crate) struct RetryableTest {
        pub keys: HashSet<Key>,
        fails: u32,
        pub attempts: u32,
    }

    impl RetryableTest {
        pub(crate) fn new(keys: Vec<Key>, fails: u32) -> Self {
            Self {
                keys: HashSet::from_iter(keys.into_iter()),
                fails,
                attempts: 0,
            }
        }
    }

    #[async_trait]
    impl RetryableStreamRequest for Arc<Mutex<RetryableTest>> {
        type Item = Key;

        async fn perform(&self, _client: Client) -> Result<Response<Self::Item>, EdenApiError> {
            let mut this = self.lock();
            let mut response = Response::empty();
            let mut entries: Vec<Result<Key, EdenApiError>> =
                this.keys.iter().map(|k| Ok(k.clone())).collect();

            this.attempts += 1;
            if this.attempts <= this.fails {
                entries[1] = Err(EdenApiError::Http(HttpClientError::BadResponse(anyhow!(
                    "fake error"
                ))));
            }
            response.entries = Box::pin(Box::new(stream::iter(entries.into_iter())));
            Ok(response)
        }

        fn received_item(&mut self, item: &Self::Item) {
            self.lock().keys.remove(item);
        }
    }

    #[test]
    fn test_retryable_trait() -> Result<()> {
        let base_url = "https://example.com".parse()?;
        let repo_name = "test_repo";

        let client = HttpClientBuilder::new()
            .repo_name(repo_name)
            .server_url(base_url)
            .max_retry_per_request(10)
            .build()?;

        let keys: Vec<Key> = vec![
            Key::new(
                RepoPathBuf::from_string("1".to_string()).unwrap(),
                HgId::null_id().clone(),
            ),
            Key::new(
                RepoPathBuf::from_string("2".to_string()).unwrap(),
                HgId::null_id().clone(),
            ),
            Key::new(
                RepoPathBuf::from_string("3".to_string()).unwrap(),
                HgId::null_id().clone(),
            ),
        ];
        let fails = 2;
        let retryable = Arc::new(Mutex::new(RetryableTest::new(keys, fails)));
        let retryable_move = retryable.clone();
        let response = block_on(retryable_move.perform_with_retries(client))?;

        let results: Vec<_> = stream_to_iter(response.entries).collect();

        assert_eq!(retryable.lock().attempts, fails + 1);
        assert_eq!(retryable.lock().keys.len(), 0);
        let results: HashSet<String> =
            HashSet::from_iter(results.into_iter().map(|k| k.unwrap().path.into_string()));
        assert!(results.contains(&"1".to_string()));
        assert!(results.contains(&"2".to_string()));
        assert!(results.contains(&"3".to_string()));

        Ok(())
    }
}
