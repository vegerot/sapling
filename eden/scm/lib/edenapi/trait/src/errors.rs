/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use std::time::Duration;

use edenapi_types::wire::WireToApiConversionError;
use edenapi_types::SaplingRemoteApiServerError;
use http::header::HeaderMap;
use http::status::StatusCode;
use http_client::HttpClientError;
use http_client::TlsError;
use http_client::TlsErrorKind;
use thiserror::Error;
use types::errors::NetworkError;

#[derive(Debug, Error)]
pub enum SaplingRemoteApiError {
    #[error("failed to serialize request: {0}")]
    RequestSerializationFailed(#[source] serde_cbor::Error),
    #[error("failed to parse response: {0}")]
    ParseResponse(String),
    #[error(transparent)]
    BadConfig(#[from] ConfigError),
    #[error(transparent)]
    Http(#[from] HttpClientError),
    #[error("server responded {status} for {url}: {message}. Headers: {headers:#?}")]
    HttpError {
        status: StatusCode,
        message: String,
        headers: HeaderMap,
        url: String,
    },
    #[error(transparent)]
    InvalidUrl(#[from] url::ParseError),
    #[error(transparent)]
    WireToApiConversionFailed(#[from] WireToApiConversionError),
    #[error(transparent)]
    ServerError(#[from] SaplingRemoteApiServerError),
    #[error("expected response, but none returned by the server")]
    NoResponse,
    #[error(transparent)]
    Other(#[from] anyhow::Error),
    #[error("not supported by the server")]
    NotSupported,
    #[error(transparent)]
    MissingCerts(#[from] auth::MissingCerts),
}

#[derive(Debug, Error)]
pub enum ConfigError {
    #[error("missing required config item: {0}")]
    Missing(String),
    #[error("invalid config item: '{0}' ({1})")]
    Invalid(String, #[source] anyhow::Error),
}

impl SaplingRemoteApiError {
    pub fn is_rate_limiting(&self) -> bool {
        use SaplingRemoteApiError::*;
        match self {
            HttpError { status, .. } => {
                if status.is_client_error() {
                    match *status {
                        StatusCode::TOO_MANY_REQUESTS => true,
                        _ => false,
                    }
                } else {
                    false
                }
            }
            _ => false,
        }
    }

    pub fn is_retryable(&self) -> bool {
        use http_client::HttpClientError::*;
        use SaplingRemoteApiError::*;
        match self {
            Http(client_error) => match client_error {
                Tls(TlsError { kind, .. }) => kind == &TlsErrorKind::RecvError,
                _ => true,
            },
            HttpError { status, .. } => {
                // 300-399
                if status.is_redirection() {
                    false
                // 400-499
                } else if status.is_client_error() {
                    match *status {
                        StatusCode::REQUEST_TIMEOUT => true,
                        StatusCode::TOO_MANY_REQUESTS => true,
                        _ => false,
                    }
                // 500-599
                } else {
                    // request could be too heavy
                    if *status == StatusCode::GATEWAY_TIMEOUT {
                        return false;
                    }
                    status.is_server_error()
                }
            }
            _ => false,
        }
    }

    pub fn retry_after(&self, attempt: usize, max: usize) -> Option<Duration> {
        if self.is_retryable() && attempt < max {
            // Retrying for a longer period of time is simply a
            // way to wait until whatever surge of traffic is happening ends.
            if self.is_rate_limiting() {
                Some(Duration::from_secs(
                    u32::pow(2, std::cmp::min(3, attempt as u32 + 1)) as u64,
                ))
            } else {
                Some(Duration::from_secs(attempt as u64 + 1))
            }
        } else {
            None
        }
    }

    // Report whether this error may be a network error. Err on the side of saying "yes".
    pub fn maybe_network_error(&self) -> bool {
        use SaplingRemoteApiError::*;

        match self {
            Http(_) | HttpError { .. } | ServerError(_) | NoResponse | Other(_) => true,

            RequestSerializationFailed(_)
            | ParseResponse(_)
            | BadConfig(_)
            | InvalidUrl(_)
            | WireToApiConversionFailed(_)
            | NotSupported
            | MissingCerts(_) => false,
        }
    }

    pub fn tag_network(self) -> anyhow::Error {
        if self.maybe_network_error() {
            NetworkError::wrap(self)
        } else {
            self.into()
        }
    }
}
