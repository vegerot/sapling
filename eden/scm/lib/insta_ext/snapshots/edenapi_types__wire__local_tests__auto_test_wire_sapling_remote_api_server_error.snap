---
source: mod.rs
expression: "WireSaplingRemoteApiServerError::arbitrary(&mut g)"

---
{"1":"Â€"}
